### PR TITLE
server: add Host Control Plane State to uservm and systemvm response

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR appends host control plane state to UserVM, domain router and systemvm details.

The host control plane state is determined by the host status and host resource state. The options are
- Enabled
- Disabled
- Maintenance
- Offline
- Unknown

Some VM actions are disallowed on UI
- Attach ISO: If host control state is Offline or Maitenance
- Detach ISO: If host control state is Offline or Maitenance
- Reinstall VM: If host control state is Offline
- Migrate VM:  If host control state is Offline
- Migrate VM with volume:  If host control state is Offline
- VM console:  If host control state is Offline

The matrix of Host control plane state can be found at
https://cwiki.apache.org/confluence/display/CLOUDSTACK/Host+Control+Plane+State+for+VMs

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test matrix

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test matrix

**Comment by borisstoyanov** [comment]
@blueorangutan package

File: api/src/test/java/com/cloud/host/ControlStateTest.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import org.junit.Test;
+
+public class ControlStateTest extends TestCase {
+
+    void verifyHostControlState(Status hostStatus, ResourceState hostResourceState, ControlState expectedControlState) {
+        Assert.assertEquals(expectedControlState, ControlState.getControlState(hostStatus, hostResourceState));
+    }
+
+    @Test
+    public void testHostControlState() {
```
Comment (DaanHoogland) [inline api/src/test/java/com/cloud/host/ControlStateTest.java:33]: should these all be in one test? (if the first fails none of the others will be tested)

File: test/integration/smoke/test_host_control_state.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+            serviceofferingid=cls.service_offering.id
+        )
+
+        cls._cleanup = [
+            cls.domain,
+            cls.account,
+            cls.vm,
+            cls.service_offering
+        ]
+        return
```
Comment (DaanHoogland) [inline test/integration/smoke/test_host_control_state.py:91]:
```suggestion
        cls._cleanup = []

        cls.domain = Domain.create(
            cls.apiclient,
            cls.services["acl"]["domain1"]
        )
        cls._cleanup.append(cls.domain)
        cls.account = Account.create(
            cls.apiclient,
            cls.services["account"],
            domainid=cls.domain.id
        )
        cls._cleanup.append(cls.account)
        cls.service_offering = ServiceOffering.create(
            cls.apiclient,
            cls.services["service_offerings"]["tiny"]
        )
        cls._cleanup.append(cls.service_offering)
        cls.vm = VirtualMachine.create(
            cls.apiclient,
            cls.services["virtual_machine"],
            templateid=cls.template.id,
            accountid=cls.account.name,
            domainid=cls.account.domainid,
            serviceofferingid=cls.service_offering.id
        )
        cls._cleanup.append(cls.vm)
```

**Review by DaanHoogland — APPROVED** [review]
clgtm, just some remarks on the tests. 
Also i wonder how we assess completeness of this enhancement. guess thourough testing ;)

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by alexandremattioli** [comment]
LGTM

**Pull request merged**